### PR TITLE
[BootstrapAdminPanel] Payment Methods Create and Update Views

### PR DESCRIPTION
**Pull request by Rafikooo**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | bootstrap-admin-panel                   |
| Bug fix?        | no                                                       |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| License         | MIT                                                          |

image

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPaymentMethodsContext.php [hunk 1]
```diff
@@ -460,15 +496,15 @@ public function iConfigureItForUsernameWithSignature($username, $signature)
...
      */
-    public function iDoNotSpecifyConfigurationPassword()
+    public function iDoNotSpecifyConfigurationPassword(): void
     {
         // Intentionally left blank to fulfill context expectation
     }
 
     /**
      * @When I configure it with test stripe gateway data
      */
```
Comment (mpysiak) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingPaymentMethodsContext.php:549]:
```suggestion
    /**
     * @When I configure it with test stripe gateway data
     *
     * @throws InvalidArgumentException
     */
```

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/translations.html.twig [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    <div class="card-header">
+        <div class="card-title">{{ 'sylius.ui.translations'|trans }}</div>
+    </div>
+
+    <div class="card-body">
+        <div class="row">
+            {% hook [main_hook, fallback_hook] with hookable_data|merge({
+                parent_main_hook: main_hook,
+                parent_fallback_hook: fallback_hook,
+            }) %}
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/translations.html.twig]:
```suggestion
            {% set body %}
                {% verbatim %}
                    <div class="row"">
                        <div class="col-12">
                            {{ form_row(translationForm.name) }}
                            {{ form_row(translationForm.description) }}
                            {{ form_row(translationForm.instructions) }}
                        </div>
                    </div>
                {% endverbatim %}
            {% endset %}
            {{ _translations.default(hookable_data.form.translations, body) }}
```

File: src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/payment_method/update.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            enabled:
+                template: '@SyliusAdmin/payment_method/form/sections/general/enabled.html.twig'
+            channels:
+                template: '@SyliusAdmin/payment_method/form/sections/general/channels.html.twig'
+
+        'sylius_admin.payment_method.update.content.form.sections.gateway_configuration':
+            type:
+                template: '@SyliusAdmin/payment_method/form/sections/gateway_configuration/type.html.twig'
+
+        'sylius_admin.payment_method.update.content.form.sections.translations':
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/payment_method/update.yaml]: There are missing hooks for `paypal` and `stripe` configuration.

File: src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php [hunk 4]
```diff
@@ -23,63 +24,96 @@ class UpdatePage extends BaseUpdatePage implements UpdatePageInterface
     use ChecksCodeImmutability;
     use Toggles;
 
+    /**
+     * @throws ElementNotFoundException
+     */
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php]:
```suggestion
    /** @throws ElementNotFoundException */
```

**Review by GSadee — COMMENTED** [review]
The rebase is needed

File: src/Sylius/Behat/Context/Ui/Admin/ManagingPaymentMethodsContext.php (hunk 1 above)
Comment (Rafikooo) [reply src/Sylius/Behat/Context/Ui/Admin/ManagingPaymentMethodsContext.php:549]: I am unsure if this method is throwing an exception.

File: src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php (hunk 4 above)
Comment (Rafikooo) [reply src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php]: No longer relevant.

File: src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/payment_method/update.yaml (hunk 3 above)
Comment (Rafikooo) [reply src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/payment_method/update.yaml]: Fixed 🔧

File: features/admin/payment/managing_payment_methods/editing_payment_method_with_paypal_express_checkout_gateway.feature [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        Then I should be notified that it has been successfully edited
+        And its gateway configuration "Signature" should be "new_signature"
+
+    @ui @no-api
+    Scenario: Changing gateway sandbox mode
+        When I want to modify the "PayPal Express Checkout" payment method
+        And I enable sandbox mode
+        And I save my changes
+        Then I should be notified that it has been successfully edited
+        And this payment method should be in sandbox mode
```
Comment (GSadee) [inline features/admin/payment/managing_payment_methods/editing_payment_method_with_paypal_express_checkout_gateway.feature:42]: Why do these scenarios have a `no-api` tag?

File: features/admin/payment/managing_payment_methods/editing_payment_method_with_stripe_checkout_gateway.feature [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    @ui @no-api
+    Scenario: Changing Stripe Checkout gateway secret key
+        When I want to modify the "Stripe Checkout" payment method
+        And I update its "Secret key" with "some_secret_key"
+        And I save my changes
+        Then I should be notified that it has been successfully edited
+        And this payment method "Secret key" should be "some_secret_key"
+
+
```
Comment (GSadee) [inline features/admin/payment/managing_payment_methods/editing_payment_method_with_stripe_checkout_gateway.feature]:
```suggestion

```

File: src/Sylius/Behat/Context/Api/Admin/ManagingPaymentMethodsContext.php [hunk 7]
```diff
@@ -751,6 +807,20 @@ public function thisPaymentMethodShouldStillBeInTheRegistry(PaymentMethodInterfa
...
+    {
+        $availableFields = ['Publishable key', 'Secret key', 'Username', 'Password', 'Signature', 'Sandbox'];
+
+        if (!in_array($field, $availableFields)) {
+            throw new \InvalidArgumentException(sprintf('There is no configuration for "%s" field.', $field));
+        }
+
+        $this->updateGatewayConfig([StringInflector::nameToLowercaseCode($field) => $value]);
+    }
+
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingPaymentMethodsContext.php]: `When` method should be before `Then` ones

File: src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php [hunk 8]
```diff
@@ -73,15 +93,24 @@ protected function getToggleableElement(): NodeElement
...
+            'enabled' => '[data-test-enabled]',
+            'factory_name' => '[data-test-factory-name]',
             'instructions' => '#sylius_payment_method_translations_%language%_instructions',
             'name' => '#sylius_payment_method_translations_en_US_name',
+            'publishable_key' => '[data-test-publishable-key]',
+            'secret_key' => '[data-test-secret-key]',
+            'username' => '[data-test-username]',
+            'password' => '[data-test-password]',
+            'signature' => '[data-test-signature]',
+            'sandbox' => '[data-test-sandbox]',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php]: Elements should be in alphabetical order

File: src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php [hunk 9]
```diff
@@ -73,15 +93,24 @@ protected function getToggleableElement(): NodeElement
...
+     * @return array<string, string>
+     */
     protected function getDefinedElements(): array
     {
         return array_merge(parent::getDefinedElements(), [
-            'channel' => '.checkbox:contains("%channel%") input',
-            'code' => '#sylius_payment_method_code',
-            'enabled' => '#sylius_payment_method_enabled',
-            'factory_name' => '#sylius_payment_method_gatewayConfig_factoryName',
+            'channel' => '[data-test-channels] .form-switch:contains("%channel%") input',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/PaymentMethod/UpdatePage.php]: Ideally, you should not use classes to define elements

File: src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/gateway_configuration.html.twig [hunk 10]
```diff
@@ -0,0 +1,24 @@
+{% set main_hook = hook_name(hookable_data.parent_main_hook, 'gateway_configuration') %}
+{% set fallback_hook = hook_name(hookable_data.parent_fallback_hook, 'gateway_configuration') %}
+
+{% set gatewayType = hookable_data.form.vars.data.gatewayConfig.factoryName %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/gateway_configuration.html.twig]:
```suggestion
{% set gateway_type = hookable_data.form.vars.data.gatewayConfig.factoryName %}
```

File: src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/translations.html.twig [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+    <div class="card-body">
+        <div class="row">
+            {% set body %}
+                {% verbatim %}
+                    <div class="row"">
+                        <div class="col-12">
+                            {{ form_row(translationForm.name) }}
+                            {{ form_row(translationForm.description) }}
+                            {{ form_row(translationForm.instructions) }}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/payment_method/form/sections/translations.html.twig]: We should split these fields into separate hooks

File: features/admin/payment/managing_payment_methods/editing_payment_method_with_paypal_express_checkout_gateway.feature (hunk 5 above)
Comment (Rafikooo) [reply features/admin/payment/managing_payment_methods/editing_payment_method_with_paypal_express_checkout_gateway.feature:42]: Due to validation, it is not possible to update a single field using the API. Therefore, I have created another test that updates all these fields at once, and this test has the `@api` tag.

File: features/admin/payment/managing_payment_methods/editing_payment_method_with_stripe_checkout_gateway.feature [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    @ui @no-api
+    Scenario: Changing Stripe Checkout gateway secret key
+        When I want to modify the "Stripe Checkout" payment method
+        And I update its "Secret key" with "some_secret_key"
+        And I save my changes
+        Then I should be notified that it has been successfully edited
+        And this payment method "Secret key" should be "some_secret_key"
+
+    @api @ui
```
Comment (NoResponseMate) [inline features/admin/payment/managing_payment_methods/editing_payment_method_with_stripe_checkout_gateway.feature:28]: We should really refrain from adding and covering any API scenarios in this branch.

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @Rafikooo!

**Pull request merged**